ls1intum/Artemis
# `Learning analytics`: Prevent circular relations between learning goals

### Checklist
#### General


#### Server
#### Client

### Motivation and Context


There are four different types of relations that can be added between competencies (RELATES: A generic relation between two learning goals, ASSUMES: The tail learning goal assumes that the student already achieved the head learning goal, EXTENDS: The tail learning goal extends the head learning goal on the same topic in more detail, MATCHES: The tail learning goal matches the head learning goal (e.g., a duplicate)).
As a first step to generating learning paths, we have to prevent circular ASSUMES and EXTENDS relations as they make it impossible to find a way to learn all these competencies and a circle also doesn't make sense semantically.

### Description

To achieve this we generate a graph with all competencies and competency relations, traverse it depth-first, mark already visited nodes, and check if neighboring nodes are already marked as visited. If that is the case there has to be a circle. We have to repeat this with every node as starting point because otherwise, we would miss nodes if the graph is not connected. If no circle was detected there is no circle.
If you now want to add a relation that will result in a cycle you get an error and the button will be disabled.
You also can't create a relation between a competency and itself or an already existing relation.

### Steps for Testing

Prerequisites:
- 1 Admin (The feature is not yet visible to instructors)

1. Log in to Artemis
2. Navigate to Course Administration
3. Select a course
4. Click on competencies
5. Create competencies and relations between them and make sure it is not possible to create a circle with EXTENDS and ASSUMES

### Review Progress






#### Code Review
#### Manual Tests


### Test Coverage





| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|-------:|-----:|
| LearningGoalService.java | 99% | ✅  |
| learning-goal-management.component.ts | 81% | ✅  |

### Screenshots



![Screenshot 2023-04-14 at 02-02-13 Competency Management](https://user-images.githubusercontent.com/44734881/231908687-00c59ee3-df86-498f-9830-0803f65c6a89.png)

![Screenshot 2023-04-14 at 02-01-37 Competency Management](https://user-images.githubusercontent.com/44734881/231908727-400abbba-7df0-4b64-8c8a-1fb4ddf5bcbb.png)
